Allow to add middleware and allow admin roles

yaroslavmo opened the pull request.

OlegLazaryev commented on src/Models/AdminUser.php:
```diff
@@ -15,9 +15,69 @@ public static function bulkCreate(array $records, array $params = [])
+     *
+     * @return array $role
+     */
+    protected static function createAdminRole($records)
```
1. Maybe it is better to name this function something like `createRestrictedAdminRole`?
2. Should this function be private?

yaroslavmo commented on src/Models/AdminRoleServicesAccessor.php:
```diff
@@ -0,0 +1,194 @@
+     */
+    private function verifyRoleId($roleId)
+    {
+        if (0 === $roleId) {
```
nope added it just in case. I couldn't reproduce so I deleted it

OlegLazaryev commented on src/Components/RestrictedAdminRoleCreator.php:
```diff
@@ -0,0 +1,300 @@
+     * @return array
+     * @throws \Exception
+     */
+    public static function createAndLinkRestrictedAdminRole(array $records)
```
Do you see any alternatives for naming the argument of this finction?
From my perspective `$records` seems a bit abstract so I am wondering whether you could make it more specific.

yaroslavmo replied:
call createRestrictedAdminRole and then call linkRestrictedAdminRole using class instance. So two separate functions. Good decision?

OlegLazaryev commented on src/Components/RestrictedAdminRoleCreator.php:
```diff
@@ -0,0 +1,300 @@
+
+        if (isset($records[0]["access_by_tabs"]) && in_array("files", $records[0]["access_by_tabs"])) {
+            array_push($userToAppToRoleByUserId, ["app_id" => App::whereName("file_manager")->first()["id"], "role_id" => $roleId]);
+        }
```
Let's discuss these two conditions real quick.

OlegLazaryev commented on src/Components/RestrictedAdminRoleCreator.php:
```diff
@@ -0,0 +1,300 @@
+            "config" => array("name" => "config", "label" => "Config", "path" => "/config"),
+            "packages" => array("name" => "package-manager", "label" => "Packages", "path" => "/package-manager"),
+            "limits" => array("name" => "limit", "label" => "Limits", "path" => "/limits")
+        );
```
This is not a good approach to keep view details in this low-level code. Let's discuss.

yaroslavmo replied:
removed them

OlegLazaryev commented on src/Components/RestrictedAdminRoleCreator.php:
```diff
@@ -0,0 +1,300 @@
+/**
+ * Services access by tabs for admin role
+ */
+class RestrictedAdminRoleCreator
```
If I am not missing something all the methods of this class are static. As it usually happens it causes the need to bypass parameters from one function to another. 

Let's discuss how we can turn this class from static to instance approach. I feel like we talked about that already but anyway.

yaroslavmo replied:
now only methods that used in tabs visibility are static. methods that create and link role require the class instance. Is that ok?

OlegLazaryev replied:
Yeah, totally. Good job.

OlegLazaryev commented on src/Models/AdminUser.php:
```diff
@@ -15,6 +16,16 @@ public static function bulkCreate(array $records, array $params = [])
 
+        $isRestrictedAdmin = isset($records[0]["is_restricted_admin"]) && $records[0]["is_restricted_admin"];
+        $accessByTabs = isset($records[0]["access_by_tabs"]) ? $records[0]["access_by_tabs"] : [];
+        if ($isRestrictedAdmin && !RestrictedAdminController::isAllTabs($accessByTabs)) {
```
This breaks MVC model approach. Model should not be aware about details from Controller level.
Please extract the common logic into a model.

yaroslavmo replied:
Moved everything to model and separate it by a comment. It required to make every method static and pass parameters between them. Luckily there are only 3 "global" parameters that need to be passed to restricted admin related functions. Then functions manage their own "local" parameters between them.

OlegLazaryev commented on src/Models/AdminUser.php:
```diff
@@ -15,6 +16,16 @@ public static function bulkCreate(array $records, array $params = [])
+        $accessByTabs = isset($records[0]["access_by_tabs"]) ? $records[0]["access_by_tabs"] : [];
+        if ($isRestrictedAdmin && !RestrictedAdminController::isAllTabs($accessByTabs)) {
+            $restrictedAdminController = new RestrictedAdminController($records[0]["email"], $accessByTabs);
+            $restrictedAdminController->createRestrictedAdminRole();
```
See the comment above.
Very bad practice to reference Controllers from inside the Model.

OlegLazaryev commented on src/Models/AdminUser.php:
```diff
@@ -50,7 +61,17 @@ public static function bulkUpdate(array $records, array $params = [])
+            $records[0]["user_to_app_to_role_by_user_id"] = $restrictedAdminController->getUserAppRoleByUserId($isRestrictedAdmin, $records[0]["id"]);
+        } elseif (!$isRestrictedAdmin) {
+            $restrictedAdminController->deleteRole($records[0]["id"]);
+        };
```
The same as above.

OlegLazaryev commented on src/Models/Role.php:
```diff
@@ -63,6 +64,19 @@ function (Role $role){
+        $model = parent::selectById($id, $options, $fields);
+
+        if ($model && self::isAccessibleTabsSpecified($options)) {
+            $model["accessible_tabs"] = RestrictedAdminController::getAccessibleTabsByRoleId($model["id"]);
```
Let's not call the controller here. Extract common logic into a model.

yaroslavmo replied:
extracted all methods, except getTabsAccessesMap. This method returns tab accesses map and I don't want to duplicate this map in the Role model. 
So still need to call AdminUser::getTabsAccessesMap() to get it.

OlegLazaryev commented on src/Models/Role.php:
```diff
@@ -63,6 +64,19 @@ function (Role $role){
+     */
+    public static function selectById($id, array $options = [], array $fields = ['*'])
+    {
+        $model = parent::selectById($id, $options, $fields);
```
Let's rename `$model` into `$role` or whatever that object represents.

OlegLazaryev commented on src/Utility/Session.php:
```diff
@@ -70,7 +70,7 @@ public static function checkServicePermission(
     public static function getServicePermissions($service, $component = null, $requestor = ServiceRequestorTypes::API)
     {
-        if (static::isSysAdmin()) {
+        if (static::isSysAdmin() && !Session::getRoleId()) {
```
`!Session::getRoleId()`
Let's extract this expression into a function and name it appropriately.

OlegLazaryev commented on src/Utility/Session.php:
```diff
@@ -279,7 +279,7 @@ public static function checkForAnyServicePermissions(
     public static function allowsServiceAccess($service, $requestor = ServiceRequestorTypes::API)
     {
-        if (static::isSysAdmin()) {
+        if (static::isSysAdmin() && !UserAppRole::whereUserId(Session::getCurrentUserId())->first()) {
```
`!UserAppRole::whereUserId(Session::getCurrentUserId())->first()`
Let's extract this expression into a function and name it appropriately.

OlegLazaryev commented on src/Models/AdminUser.php:
```diff
@@ -122,4 +152,557 @@ protected static function fixRecords(array $records)
+    private static function getServiceIdByName(string $name)
+    {
+        return Service::whereName($name)->get(["id"])->first()["id"];
+    }
```
Can we move this function into a Service model?

yaroslavmo replied:
No need to do so. I found out that there is a ServiceManager which has its own getServiceIdByName method which works fine. So I used it

OlegLazaryev commented on src/Models/AdminUser.php:
```diff
@@ -122,4 +152,557 @@ protected static function fixRecords(array $records)
+     * @param $email
+     * @return bool
+     */
+    public static function getAdmin($email)
```
Shouldn't we rename this to `getUserByEmail`?
And probably move to a User model.

yaroslavmo replied:
I renamed it to `getAdminByEmail`. It may stay here as this class extends User and the User here is always admin.

yaroslavmo replied:
I just misused User:: here it should have been self::

OlegLazaryev commented on src/Models/AdminUser.php:
```diff
@@ -122,4 +152,557 @@ protected static function fixRecords(array $records)
+    }
+
+    /**
+     * Get userId.
```
This comment should be probably renamed too.

OlegLazaryev commented on src/Models/AdminUser.php:
```diff
@@ -122,4 +152,557 @@ protected static function fixRecords(array $records)
+
+        foreach ($tabs as $tab) {
+            $roleServiceAccess = self::getTabServiceAccesses($roleServiceAccess, $tab, $withDefault, $withTabs, $unique);
+        }
```
Am I correct that this loop always returns the result of last step?
If we need only a single result maybe it is better to reference it directly like `$tabs[0]`

yaroslavmo replied:
Actually no. because of `**$roleServiceAccess** = self::getTabServiceAccesses(**$roleServiceAccess**, $tab, $withDefault, $withTabs, $unique);` it takes the result array and pushes to it more values if they are not already in it.

yaroslavmo commented:
> @yaroslavmo I can see a lot of new functionality added to `src/Models/AdminUser.php`
> Since all this functionality is specific to Restricted Admin feature it makes sense to research on how to simply extract it into a separate class in a separate file.

I must have confused you. I had `restrictedAdminRoleCreator` class which had all logic related to restricted admins -> after I added update functionality I renamed it to `restrictedAdminController` because it then wasn't only creating restrictedAdmin's roles but also updates them -> after review we agreed that this logic should not belong to the Controller but to the Model -> I got rid of it -> and now, as I understand, we're coming back to it?

OlegLazaryev commented:
> > @yaroslavmo I can see a lot of new functionality added to `src/Models/AdminUser.php`
> > Since all this functionality is specific to Restricted Admin feature it makes sense to research on how to simply extract it into a separate class in a separate file.
> 
> @OlegLazaryev I must have confused you. I had `restrictedAdminRoleCreator` class which had all logic related to restricted admins -> after I added update functionality I renamed it to `restrictedAdminController` because it then wasn't only creating restrictedAdmin's roles but also updating them -> after review we agreed that this logic should not belong to the Controller but to the Model -> I got rid of it -> and now, as I understand, we're coming back to this approach? But how to call this class then?

Yeah, you are correct. I was confused by the naming `Controller`.

OlegLazaryev commented on src/Utility/Session.php:
```diff
@@ -1003,7 +1002,7 @@ public static function getApiKey()
     public static function isSysAdmin()
     {
-        return boolval(session('user.is_sys_admin'));
+        return boolval(session('user.is_sys_admin')) && !self::isRestrictedAdmin();
```
@yaroslavmo could you please describe shortly which particular business case this change is supposed to handle?
I am interested in what are the consequences if won't introduce this change into `isSysAdmin` function.

yaroslavmo replied:
If won't introduce this change, every admin will have access to all services and all request methods.
We don't want that because for the RA feature we want roles to take place in defining admin's permissions.

yaroslavmo commented:
Deprecated. Already in master
